src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -1474,25 +1507,40 @@ public void exitSegmentMode() throws IOException {
          * If it is too large, lots of memory is used in RAM.
          */
         private boolean shouldFlush() {
+            // unbuffered mode always flush data.
+            if (unbufferedMode) return true;
             // return true if data in segmentBuffer has been extended.
             return segmentWritten > SEGMENT_BUFFER_SIZE;
         }
 
         /**
-         * Writes the write segment header into internal buffer.
+         * Writes the segment header with given data size.
          */
-        private void writeSegmentHeader() {
-            assert segmentWritten == 0;
-            segmentBuffer[segmentWritten++] = (byte)HPROF_HEAP_DUMP_SEGMENT;
-            writeInteger(0);
-            // segment size, write dummy length of 0 and we'll fix it later.
+        private void writeSegmentHeader(int size) throws IOException {
+            assert segmentWritten == 0 : "initializing non empty segment";
+            byte flag = (byte)HPROF_HEAP_DUMP_SEGMENT;
+            if (unbufferedMode) {
+                super.write(flag);
+            } else {
+                segmentBuffer[segmentWritten++] = flag;
+            }
             writeInteger(0);
+            writeInteger(size);
```

Comment:
You've added `writeInteger(size)` here, but I'm not sure where this used to be done.

Reply:
There used to be 2 writeInterger(0), and the second one is write the segment size with dummy value 0. I will add a comment here. Thanks

Reply:
Ok, I missed that. I think where a comment is needed is with the first `writerInteger(0)`. It's not clear to me what that one is for.

Reply:
OK, I will add it.  it is a dummy value for `timestamp` slot.And seems it is never fill with real time stamp in current implementation.